File: jump_game_1_bfs.py
# Time Complexity : O(N^M) where N elements in the array and M is max (jump capacity/strength) number in the array
# Space Complexity : O(N) where N is number of elements in the array. Used for hashset and queue
# Did this code successfully run on Leetcode : No (Time Limit Exceeded)
# Any problem you faced while coding this : No

#BFS Solution
from collections import deque
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        end = len(nums) - 1
        if end == 0: return True
        hashset = set()
        q = deque()
        q.appendleft(0)
        hashset.add(0)
        
        
        while len(q) != 0:
            index = q.pop()
            maxJump = nums[index]
            
            for i in range(1, maxJump + 1):
                nextIndex = index + i
                if nextIndex >= end:
                    return True
                
                if nextIndex not in hashset:
                    q.appendleft(nextIndex)
                    hashset.add(nextIndex)
                    
        return False

File: JumpGame1BFS.java
// Time Complexity : O(k^n) -> k is the average number of jumps that can be made and n it the nums length
// Space Complexity : O(k^n)
// Did this code successfully run on Leetcode : Yes

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;

public class JumpGame1BFS {
    public boolean canJump(int[] nums) {
        if(nums.length <=1) return true;
        Queue<Integer> q = new LinkedList<>();
        HashSet<Integer> visited = new HashSet<>();
        q.add(0);

        while(!q.isEmpty()){
            int currIdx = q.poll();
            for(int i=1; i<= nums[currIdx]; i++)
            {
                int newIdx = currIdx + i;
                if(newIdx == nums.length-1) return true;
                if(!visited.contains(newIdx))
                {
                    visited.add(newIdx);
                    q.add(newIdx);
                }
            }
        }
        return false;
    }
}
